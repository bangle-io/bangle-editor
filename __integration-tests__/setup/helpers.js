const prettier = require('prettier');
const os = require('os');

const pmRoot = '#pm-root';
const ctrlKey = os.platform() === 'darwin' ? 'Meta' : 'Control';

module.exports = {
  mountEditor,
  getEditorState,
  pressRight,
  pressLeft,
  debug,
  ctrlKey,
  pmRoot,
  getDoc,
  sleep,
};

function frmt(doc) {
  return prettier.format(doc.toString(), {
    semi: false,
    parser: 'babel',
    printWidth: 36,
    singleQuote: true,
  });
}

async function mountEditor(page, props) {
  await page.waitForSelector(pmRoot, { timeout: 500 });
  await page.waitForSelector('.ProseMirror', { timeout: 500 });
}

async function getEditorState(page) {
  return page.evaluate(() => {
    return window.editor.view.state.toJSON();
  });
}

async function pressRight() {
  await page.keyboard.press('ArrowRight');
}

async function pressLeft() {
  await page.keyboard.press('ArrowLeft');
}

function debug() {
  return jestPuppeteer.debug();
}

function sleep(t = 20) {
  return new Promise((res) => setTimeout(res, t));
}

async function getDoc(page) {
  return page
    .evaluate(() => {
      return window.editor.view.state.doc.toString();
    })
    .then(frmt);
}
