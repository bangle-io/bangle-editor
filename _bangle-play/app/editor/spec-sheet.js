import '@banglejs/core/style.css';
import '@banglejs/tooltip/style.css';

import * as collab from '@banglejs/collab/client/collab-extension';
import { emoji } from '@banglejs/emoji/index';
import '@banglejs/emoji/emoji.css';
import '@banglejs/react-menu/style.css';
import '@banglejs/markdown-front-matter/style.css';
import './extensions-override.css';
import { trailingNode } from '@banglejs/trailing-node';
import { timestamp } from '@banglejs/timestamp';
import { SpecRegistry } from '@banglejs/core/spec-registry';
import stopwatch from '@banglejs/react-stopwatch';
import sticker from '@banglejs/react-sticker';
import { emojiSuggestMenu } from '@banglejs/react-menu';
import { markdownFrontMatter } from '@banglejs/markdown-front-matter';
import {
  bold,
  code,
  italic,
  strike,
  link,
  underline,
  doc,
  text,
  paragraph,
  blockquote,
  bulletList,
  codeBlock,
  hardBreak,
  heading,
  horizontalRule,
  listItem,
  orderedList,
  todoItem,
  todoList,
  image,
} from '@banglejs/core/components/index';

export const specRegistry = new SpecRegistry([
  doc.spec({ content: 'frontMatter? block+' }),
  text.spec(),
  paragraph.spec(),
  blockquote.spec(),
  bulletList.spec(),
  codeBlock.spec(),
  hardBreak.spec(),
  heading.spec(),
  horizontalRule.spec(),
  listItem.spec(),
  orderedList.spec(),
  todoItem.spec(),
  todoList.spec(),
  image.spec(),
  bold.spec(),
  code.spec(),
  italic.spec(),
  strike.spec(),
  link.spec(),
  underline.spec(),
  collab.spec(),
  emoji.spec(),
  emojiSuggestMenu.spec(),
  stopwatch.spec(),
  trailingNode.spec(),
  timestamp.spec(),
  sticker.spec(),
  markdownFrontMatter.spec(),
]);
