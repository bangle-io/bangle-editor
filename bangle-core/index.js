import * as logging from './utils/logging';
import browser, { isChromeWithSelectionBug } from './utils/browser';
export * from './components/index';
export * from './editor';
export { logging, browser, isChromeWithSelectionBug };
export * from './pm-plugin';
export * from './node-view';
export * from './pm-model';
