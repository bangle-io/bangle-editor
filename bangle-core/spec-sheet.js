import { Schema } from 'prosemirror-model';
import { coreSpec } from './components/index';
import { bangleWarn, recursiveFlat } from './utils/js-utils';

export class SpecSheet {
  constructor(rawSpecs = coreSpec()) {
    this._spec = recursiveFlat(rawSpecs);

    this._spec.forEach(validateSpec);

    if (this._spec.length !== new Set(this._spec.map((r) => r.name)).size) {
      bangleWarn(
        'The specSheet has one or more specs with the same name',
        this._spec,
      );
      throw new Error('Duplicate spec error, please check your specSheet');
    }
    this._schema = createSchema(this._spec);
    this._options = Object.fromEntries(
      this._spec
        .filter((spec) => spec.options)
        .map((spec) => [spec.name, spec.options]),
    );
  }

  get spec() {
    return this._spec;
  }

  get schema() {
    return this._schema;
  }

  get options() {
    return this._options;
  }
}

function createSchema(specSheet) {
  let nodes = [];
  let marks = [];
  let topNode;
  for (const spec of specSheet) {
    if (spec.type === 'node') {
      nodes.push([spec.name, spec.schema]);
    } else if (spec.type === 'mark') {
      marks.push([spec.name, spec.schema]);
    } else if (spec.type === 'component') {
    } else {
      throw new Error(spec.name + ' unknown type: ' + spec.type);
    }
    if (spec.topNode === true) {
      topNode = spec.name;
    }
  }

  return new Schema({
    topNode,
    nodes: Object.fromEntries(nodes),
    marks: Object.fromEntries(marks),
  });
}

function validateSpec(spec) {
  if (!spec.name) {
    bangleWarn("The spec didn't have a name field", spec);
    throw new Error('Spec must have a name');
  }
  if (!['node', 'mark', 'component'].includes(spec.type)) {
    bangleWarn('The spec must be of type node, mark or component ', spec);
    throw new Error('Invalid spec type');
  }
  if (['node', 'mark'].includes(spec.type) && !spec.schema) {
    bangleWarn(
      "The spec of type 'mark' or 'node' must have a schema field",
      spec,
    );
    throw new Error('Invalid spec schema');
  }
}
