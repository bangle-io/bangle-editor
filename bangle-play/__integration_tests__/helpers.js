const os = require('os');
const prettier = require('prettier');
const { uuid } = require('bangle-core/utils/js-utils');
const PM_ID = '.ProseMirror';

const ctrlKey = os.platform() === 'darwin' ? 'Meta' : 'Control';
const EDITOR_ID = `bangle-play`;
const EDITOR_SELECTOR = `[id^='${EDITOR_ID}']`;

module.exports = {
  mountEditor,
  getEditorState,
  ctrlKey,
  getDoc,
  sleep,
  EDITOR_SELECTOR,
  uniqDatabaseUrl,
  PM_ID,
};

function uniqDatabaseUrl() {
  return `http://localhost:4444?database=databse-${uuid(2)}`;
}

function frmt(doc) {
  return prettier.format(doc.toString(), {
    semi: false,
    parser: 'babel',
    printWidth: 36,
    singleQuote: true,
  });
}
async function mountEditor(page, props) {
  await page.waitForSelector(EDITOR_SELECTOR);
  await page.waitForSelector('.ProseMirror', { timeout: 1500 });
  await page.click(EDITOR_SELECTOR);
  // let the collab  settle down
}

async function getEditorState(page) {
  return page.evaluate(() => {
    return window.editor.view.state.toJSON();
  });
}

async function getDoc(page) {
  return page
    .evaluate(() => {
      return window.editor.view.state.doc.toString();
    })
    .then(frmt);
}

function sleep(t = 20) {
  return new Promise((res) => setTimeout(res, t));
}
