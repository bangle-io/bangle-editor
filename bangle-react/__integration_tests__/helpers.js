const ctrlKey = process.env.CI ? 'Control' : 'Meta';

module.exports = {
  url: `http://localhost:4002`,
  mountEditor,
  getEditorState,
  pressRight,
  pressLeft,
  debug,
};

async function mountEditor(page, props) {
  // await page.waitForSelector('#pm-root', { timeout: 500 });
  await page.waitForSelector('#pm-root', { timeout: 500 });
  await page.waitForSelector('.Prosemirror', { timeout: 500 });
}

async function getEditorState(page) {
  return page.evaluate(() => {
    return window.editor.view.state.toJSON();
  });
}

async function pressRight() {
  await page.keyboard.press('ArrowRight');
}

async function pressLeft() {
  await page.keyboard.press('ArrowLeft');
}

async function debug() {
  return jestPuppeteer.debug();
}
