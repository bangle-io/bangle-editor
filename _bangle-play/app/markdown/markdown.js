import * as markdown from '@banglejs/markdown/index';
import { defaultMarkdownItTokenizer } from '@banglejs/markdown/index';
import { emojiMarkdownItPlugin } from '@banglejs/emoji/index';
import { specRegistry } from 'bangle-play/app/editor/spec-sheet';
import { frontMatterMarkdownItPlugin } from '@banglejs/markdown-front-matter';

const parser = markdown.markdownParser(
  specRegistry,
  defaultMarkdownItTokenizer
    .use(emojiMarkdownItPlugin)
    .use(frontMatterMarkdownItPlugin),
);

const serializer = markdown.markdownSerializer(specRegistry);

export const markdownParser = (markdownStr) => parser.parse(markdownStr);
export const markdownSerializer = (doc) => serializer.serialize(doc);
