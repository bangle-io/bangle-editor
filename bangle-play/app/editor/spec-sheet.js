import 'bangle-core/style.css';
import { coreSpec } from 'bangle-core/components';
import * as collab from 'bangle-plugins/collab/client/collab-extension';
import { emoji, emojiInlineSuggest } from 'bangle-plugins/emoji/index';
import 'bangle-plugins/emoji/emoji.css';
// import 'bangle-react/inline-menu/inline-menu.css';
import 'bangle-react/menu/style.css';

import { config } from './config';
import { trailingNode } from 'bangle-plugins/trailing-node/index';
import { timestamp } from 'bangle-plugins/timestamp/index';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { dino } from 'bangle-react/dino';
import { stopwatch } from 'bangle-react/stopwatch';

export const specSheet = new SpecSheet([
  ...coreSpec({ heading: { levels: config.headingLevels } }),
  collab.spec(),
  emoji.spec(),
  emojiInlineSuggest.spec({ markName: 'emoji_inline_suggest', trigger: ':' }),
  stopwatch.spec(),
  trailingNode.spec(),
  timestamp.spec(),
  dino.spec(),
]);
