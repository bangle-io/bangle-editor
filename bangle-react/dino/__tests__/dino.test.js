/**
 * @jest-environment jsdom
 */
/** @jsx pjsx */
import { reactTestEditor, pjsx } from '../../__tests__/helpers/index';
import { markdownSerializer } from 'bangle-plugins/markdown/index';
import { SpecRegistry } from 'bangle-core/spec-sheet';
import { dino } from '../index';
import {
  defaultPlugins,
  defaultSpecs,
} from 'bangle-core/test-helpers/default-components';

const specRegistry = new SpecRegistry([...defaultSpecs(), dino.spec()]);
const plugins = [...defaultPlugins(), dino.plugins()];

const renderNodeViews = jest.fn(({ node, ...args }) => {
  if (node.type.name === 'dino') {
    return <dino.Dino node={node} {...args} />;
  }
  throw new Error('Unknown node');
});
const testEditor = reactTestEditor({
  specRegistry,
  plugins,
  renderNodeViews,
});

test('Rendering works', async () => {
  Date.now = jest.fn(() => 0);
  const { container, view } = await testEditor(
    <doc>
      <para>
        foo[]bar
        <dino />
      </para>
    </doc>,
  );

  expect(view.state).toEqualDocAndSelection(
    <doc>
      <para>
        foo[]bar
        <dino />
      </para>
    </doc>,
  );
  expect(container.querySelector(`.bangle-dino`)).toMatchSnapshot();
});

test('Rendering works with different type of dino', async () => {
  Date.now = jest.fn(() => 0);
  const { container, view } = await testEditor(
    <doc>
      <para>
        foo[]bar
        <dino data-dinokind="stegosaurus" />
      </para>
    </doc>,
  );

  expect(view.state).toEqualDocAndSelection(
    <doc>
      <para>
        foo[]bar
        <dino data-dinokind="stegosaurus" />
      </para>
    </doc>,
  );
  expect(container.querySelector(`.bangle-dino`)).toMatchSnapshot();
});

describe('markdown', () => {
  const serialize = async (doc) => {
    let content = doc;
    if (typeof doc === 'function') {
      content = doc(specRegistry.schema);
    }
    return markdownSerializer(specRegistry).serialize(content);
  };

  test('markdown serialization', async () => {
    const md = await serialize(
      <doc>
        <para>
          hello world
          <dino data-dinokind="stegosaurus" />
        </para>
      </doc>,
    );
    expect(md).toMatchInlineSnapshot(`"hello worlddino"`);
  });
});
