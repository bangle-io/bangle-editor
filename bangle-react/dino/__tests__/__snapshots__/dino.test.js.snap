// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering works 1`] = `
<span
  class="bangle-dino "
>
  <img
    alt="brontosaurus"
    src="test-file-stub"
    style="display: inline; height: 24px; vertical-align: bottom; border: 1px solid #0ae; border-radius: 4px; background: rgb(221, 246, 255);"
  />
</span>
`;

exports[`Rendering works with different type of dino 1`] = `
<span
  class="bangle-dino "
>
  <img
    alt="stegosaurus"
    src="test-file-stub"
    style="display: inline; height: 24px; vertical-align: bottom; border: 1px solid #0ae; border-radius: 4px; background: rgb(221, 246, 255);"
  />
</span>
`;
